File: Binary_Search/1901.Find_a_Peak_Element_II/C++/1901.Find_a_Peak_Element_II.cpp
class Solution {
public:
    vector<int> findPeakGrid(vector<vector<int>> & mat) 
    {
        //===========================================//
        // Input type:                               //
        //   - vector<vector<int>> ref. (mat vector) //
        // Return type:                              //
        //   - vector<int> (result vector)           //
        //===========================================//

        /*Initialize*/
        ///// Length of rows, columns /////
        int len_rows = mat.size(), len_cols = (mat[0]).size();

        ///// Record pointers (i.e. middle, left, right) /////
        int record_mid_ptr, record_left_ptr = 0, record_right_ptr = (len_cols - 1);

        ///// Fixed row index /////
        int fixed_row_idx = 0;

        ///// Record flags (i.e left, right) /////
        bool record_lflag = false, record_rflag = false;


        /*Binary-search based loop traversal*/
        while (record_left_ptr <= record_right_ptr)
        {
            record_mid_ptr = ((record_left_ptr + record_right_ptr) / 2); //Record middle pointer

            for (int record_row_idx = 0; (record_row_idx < len_rows); (record_row_idx++)) //Rows
            {
                ///// Check if the current indexed-value is maximun or not /////
                if ((mat[record_row_idx])[record_mid_ptr] > (mat[fixed_row_idx])[record_mid_ptr]) { fixed_row_idx = record_row_idx; } //Keep updating, overwriting
                else { ; }

            } //Rows

            ///// Check if the current right-half value matched conditions or not /////
            if (((record_mid_ptr + 1) >= len_cols) || (((mat[fixed_row_idx])[(record_mid_ptr)] > (mat[fixed_row_idx])[(record_mid_ptr + 1)])))
            {
                record_rflag |= true; //Keep updating, overwriting
            }
            else { ; }

            ///// Check if the current left-half value matched conditions or not /////
            if (((record_mid_ptr - 1) < 0) || (((mat[fixed_row_idx])[record_mid_ptr] > (mat[fixed_row_idx])[(record_mid_ptr - 1)])))
            {
                record_lflag |= true; //Keep updating, overwriting
            }
            else { ; }

            ///// Check if the current left, right-half value matched conditions or not /////
            if ((record_lflag && record_rflag)) { return {fixed_row_idx, record_mid_ptr}; }

            else if (! record_lflag) { record_right_ptr = (record_mid_ptr - 1); } //Keep updating, overwriting

            else if (! record_rflag) { record_left_ptr = (record_mid_ptr + 1); } //Keep updating, overwriting

            ///// Check if the current flag needs to be reset or not /////
            if (record_lflag) { record_lflag &= false; } //Keep updating, reset
            else  { ; }

            if (record_rflag) { record_rflag &= false; } //Keep updating, reset
            else  { ; }
        }

        return {(-1), (-1)};
    }
};

File: Binary_Search/1901.Find_a_Peak_Element_II/Python/1901.Find_a_Peak_Element_II.py
class Solution(object):
    def findPeakGrid(self, mat):
        """
        :type mat: List[List[int]]
        :rtype: List[int]
        """
        #======================================#
        # Binary-search based traversal method #
        #======================================#

        ############
        #Initialize
        ##### Length of rows, coumns #####
        len_rows, len_cols = len(mat), len(mat[0])

        ##### Record pointers (i.e. left, right) #####
        record_left_ptr, record_right_ptr = 0, (len_cols - 1)

        ##### Record flags (i.e. left, right) #####
        record_lflag, record_rflag = False, False

        ##### Fixed row index #####
        fixed_row_idx = 0


        ####################################
        #Binary-search based loop traversal
        while (record_left_ptr <= record_right_ptr):
            record_mid_ptr = ((record_left_ptr + record_right_ptr) // 2) #Record middle pointer

            for record_row_idx in range(len_rows):

                ##### Check if the current indexed-value is maximun or not #####
                if ((mat[record_row_idx])[record_mid_ptr] > (mat[fixed_row_idx])[record_mid_ptr]):
                    fixed_row_idx = record_row_idx #Keep updating/overwriting
                else:
                    pass

            ##### Check if the current left-half value matched conditions or not #####
            if (((record_mid_ptr - 1) < 0) or ((mat[fixed_row_idx])[(record_mid_ptr - 1)] < (mat[fixed_row_idx])[record_mid_ptr])):
                record_lflag |= True #Keep updating/overwriting
            else:
                pass

            if (((record_mid_ptr + 1) >= len_cols) or ((mat[fixed_row_idx])[(record_mid_ptr + 1)] < (mat[fixed_row_idx])[record_mid_ptr])):
                record_rflag |= True #Keep updating/overwriting
            else:
                pass

            ##### Check if the current left, right-half value matched conditions or not #####
            if (record_lflag and record_rflag):
                return [fixed_row_idx, record_mid_ptr]

            if (not record_lflag):
                record_right_ptr = (record_mid_ptr - 1) #Keep updating/overwriting
            
            elif (not record_rflag):
                record_left_ptr = (record_mid_ptr + 1) #Keep updating/overwriting

            ##### Check if the current flag needs to be reset or not #####
            if (record_lflag):
                record_lflag &= False #Keep updating/reset
            else:
                pass

            if (record_rflag):
                record_rflag &= False #Keep updating/reset
            else:
                pass

        return [(-1), (-1)]
